### PR TITLE
Cache Before and Cache After

### PR DESCRIPTION
**Pull request by rdspring1**
Add [cache_read](https://docs.tvm.ai/api/python/schedule.html#tvm.schedule.Schedule.cache_read) and [cache_write](https://docs.tvm.ai/api/python/schedule.html#tvm.schedule.Schedule.cache_write) support.

- [X] Rename cache_read and cache_write to cache_before and cache_after respectively
- [X] Insert new TensorView Reader before current TensorView 
 e.g. For `TV2 = TV1.cache_before();` insert `TV2` before `TV1`
- [X] Insert new TensorView Reader after current TensorView 
 e.g. For `TV2 = TV1.cache_after();` insert `TV2` after `TV1`
- [x] Initial ComputeAt support - Call `TV2->computeAt(TV1, axis)` automatically if there exists a `computeAt` on `TV1`
- [x] Broadcast / Reduction support + Automatic ComputeAt replay

### CODE REVIEW

File: test/cpp/jit/test_gpu.cpp [hunk 1]
```diff
@@ -4605,6 +4605,330 @@ void testGPU_FusionReductionScheduler() {
...
+  fusion->addOutput(tv2);
+  // Before: TV1 = TV0 + 1
+  // After: TV1 = TV2; TV2 = TV0 + 1
+  // Algorithm
+
+  constexpr int BSX = 32;
+  tv2->split(-1, BSX);
+  tv0->computeAt(tv2, -1);
+
+  // cache_after automatically applies ComputeAt to the cache TensorView
```
Comment (naoyam) [inline test/cpp/jit/test_gpu.cpp]: `cache_before`?

File: test/cpp/jit/test_gpu.cpp [hunk 2]
```diff
@@ -4605,6 +4605,330 @@ void testGPU_FusionReductionScheduler() {
...
+  Fusion* fusion = prog.fusion();
+  FusionGuard fg(fusion);
+
+  TensorView* tv0 = makeDummyTensor(2);
+  TensorView* tv1 = add(tv0, new Float(1.0));
+  TensorView* tv2 = mul(tv1, new Float(3.0));
+  fusion->addInput(tv0);
+  fusion->addOutput(tv2);
+  // Before: TV1 = TV0 + 1
+  // After: TV1 = TV2; TV2 = TV0 + 1
```
Comment (naoyam) [inline test/cpp/jit/test_gpu.cpp]: Are these comments correct?

File: test/cpp/jit/test_gpu.cpp (hunk 1 above)
Comment (rdspring1) [reply test/cpp/jit/test_gpu.cpp]: Yes

File: test/cpp/jit/test_gpu.cpp (hunk 2 above)
Comment (rdspring1) [reply test/cpp/jit/test_gpu.cpp]: I changed the test but forgot to update the comments.

Comment (rdspring1) [reply test/cpp/jit/test_gpu.cpp]: ` // Before: TV2 = TV1 * 3
  // After:  TV3 = TV1 * 3;
  //         TV2 = TV3;
  // Algorithm`

Comment (naoyam) [reply test/cpp/jit/test_gpu.cpp]: Looks like there are also other inconsistent comments. Can you go through the other tests as well? It would be very helpful to understand what is really intended to do.

Comment (naoyam) [reply test/cpp/jit/test_gpu.cpp]: Example: https://github.com/csarofeen/pytorch/pull/166/files#diff-538e381bc787011cc53c15caf54e126dR4678

Comment (rdspring1) [reply test/cpp/jit/test_gpu.cpp]: The comments for `testGPU_FusionCacheAfter` should be correct. I just reviewed them.

Comment (naoyam) [reply test/cpp/jit/test_gpu.cpp]: Ah, I see. I was just wondering why it didn't have `tv2`.

File: test/cpp/jit/test_gpu.cpp [hunk 3]
```diff
@@ -4817,13 +4817,13 @@ void testGPU_FusionCacheBcast() {
   TensorView* tv5 = tv0->cache_after();
 
   // Case 2
-  // TensorView* tv5 = tv1->cache_before();
+  TensorView* tv7 = tv1->cache_before();
```
Comment (naoyam) [inline test/cpp/jit/test_gpu.cpp]: Use a consecutive number? (`tv6`)

File: torch/csrc/jit/codegen/cuda/ir_interface_nodes.h [hunk 4]
```diff
@@ -300,6 +300,16 @@ class TORCH_CUDA_API TensorView : public Val {
...
+  // Create a TensorView before the original tensor. A common use case is to
+  // read original tensor into shared memory or local registers. Analogous to
+  // TVM Cache_Write
+  TensorView* cache_before();
+
+  // Create a TensorView after the original tensor. A common use case is to
+  // write results into shared memory or local registers before moving to global
+  // memory. Analogous to TVM Cache_Read
+  TensorView* cache_after();
+
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/ir_interface_nodes.h]: Are these comments correct? Shouldn't they be swapped?

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 5]
```diff
@@ -314,6 +314,322 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
   return producer;
 }
 
+// Create a TensorView before the original tensor. A common use case is to read
+// original tensor into shared memory or local registers.
+// Analogous to TVM Cache_Write
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: This sounds like `cache_after`.

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 6]
```diff
@@ -314,6 +314,322 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
...
+        }
+      }
+    }
+  }
+
+  return producer;
+}
+
+// Create a TensorView after the original tensor. A common use case is to write
+// results into shared memory or local registers before moving to global memory.
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: This sounds like `cache_before`.

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 7]
```diff
@@ -314,6 +314,322 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
...
+        }
+      }
+    }
+  }
+
+  return consumer;
+}
+
+// Create New Expr given consumer - [output of the expression]
+struct CreateExprConsumer : public OptInDispatch {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp:493]: Can you put this inside an anonymous namespace?

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 8]
```diff
@@ -314,6 +314,322 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
...
+  void handle(BroadcastOp* broadcast_expr) final {
+    new BroadcastOp(consumer_, broadcast_expr->in());
+  }
+
+ private:
+  TensorView* consumer_;
+};
+
+// Create New Expr given producer - [an input for the expression]
+struct CreateExprProducer : public OptInDispatch {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp:545]: This one should also be in an anonymous namespace?

File: torch/csrc/jit/codegen/cuda/transform_replay.cpp [hunk 9]
```diff
@@ -430,7 +430,7 @@ std::pair<TensorDomain*, unsigned int> TransformReplay::replayCasP(
     size_t itc = 0, itp = 0;
     while (itc < consumer_root.size() || itp < producer_root.size()) {
       if (itc < consumer_root.size() && consumer_root[itc]->isBroadcast() &&
-          (itp > producer_root.size() || !producer_root[itp]->isBroadcast())) {
+          (itp >= producer_root.size() || !producer_root[itp]->isBroadcast())) {
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/transform_replay.cpp:433]: Is this related to `cache_before/after`? Or, does this fix an existing bug?

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 10]
```diff
@@ -314,6 +314,322 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
...
+  CreateExprConsumer::create(expr, consumer);
+}
+
+// Create New Expr given producer - [an input for the expression]
+void TensorView::createExprProducer(
+    Expr* expr,
+    TensorView* current,
+    TensorView* producer) {
+  CreateExprProducer::create(expr, current, producer);
+}
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: Can you please add more detailed comments about these two functions?

File: torch/csrc/jit/codegen/cuda/ir_interface_nodes.h (hunk 4 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/ir_interface_nodes.h]: I switched them back.

File: test/cpp/jit/test_gpu.cpp (hunk 3 above)
Comment (rdspring1) [reply test/cpp/jit/test_gpu.cpp]: Must have miscounted

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 10 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp]:
```
  // In Cache Before, for the origin expr of the original tensor,
  // we create a new operation where the original tensor is replaced 
  // with the new cache tensor. This function creates a new expr
  // given the consumer, the output of the expression.
  void createExprConsumer(Expr* expr, TensorView* consumer);

  // In Cache After, for all the uses of the original tensor, we create
  // a new operation where the original tensor is replaced with the new
  // cache tensor. This function creates a new expr given a producer, 
  // an input for the expression.
  void createExprProducer(Expr* expr, TensorView* current, TensorView* producer);
```

**Comment by naoyam** [comment]
@rdspring1 I added a failing test case to this branch (f2a90dc6330d6183ffc085bb19de572abb071d91). The generated code is:

```
__global__ void CUDAGeneratedKernel(Tensor<float, 1> T0, Tensor<float, 1> T2, Tensor<float, 1> T4){
  for(size_t i26 = 0; i26 < T4.size[0]; ++i26 ) {
    float T5[1];
    T5[ 0 ]
       = T0[ ( i26 * T0.stride[0] ) ];
    float T3[1];
    T3[ 0 ]
       = T5[ 0 ]
       + float(1);
    T4[ ( i26 * T4.stride[0] ) ]
       = T3[ 0 ]
       + float(2);
  }
  for(size_t i29 = 0; i29 < T4.size[0]; ++i29 ) {
    float T1[1];
    T1[ 0 ]
       = T5[ 0 ]
       + float(1);
    T2[ ( i29 * T2.stride[0] ) ]
       = T1[ 0 ]
       + float(2);
  }
}
```
As you can see, the cache of `tv0` is created but only visible in the first loop nest. 

This seems to be similar to the softmax replication issue. I'm not quite sure who is to blame in this case. Is this considered a limitation of the fuser? Or, is it something the user should not do?

File: torch/csrc/jit/codegen/cuda/transform_replay.cpp (hunk 9 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/transform_replay.cpp:433]: This is a bug I found while working on cache_before/after. If `itp == producer_root.size()`, then `producer_root[itp]` goes out of bounds.

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 7 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp:493]: Done

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 8 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp:545]: Done

**Comment by rdspring1** [comment]
I fixed the failing test case by using cache_before twice instead of a single cache_after.
```  
  // Passes
  auto tv5 = tv1->cache_before();
  auto tv6 = tv3->cache_before();

  // Fails because tensor must be recomputed twice
  // auto tv7 = tv0->cache_after();
```

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp [hunk 11]
```diff
@@ -1027,7 +1027,6 @@ Allocate::Allocate(Val* _val, Val* _size)
         ".");
   }
   addInput(_size);
-  addInput(_val);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/ir_nodes.cpp:1030]: Why?

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 12]
```diff
@@ -314,6 +314,331 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
   return producer;
 }
 
+// Create a TensorView before the original tensor. A common use case is to write
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: no need to duplicate the function comments (for both declaration and definition) -> comments on the declaration (interface) only

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 13]
```diff
@@ -314,6 +314,331 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
...
+
+  // Set domain of producer - No Change
+  TensorView* producer = this;
+
+  // Insert consumer - Cache_After (CA) - after this TV.
+  // Before: This TV -> [Use Op] -> Next TV
+  // After:  This TV -> [Set Op] -> New CA TV -> [Use Op] -> Next TV
+
+  // Expr* consumer_uses =
+  std::unordered_set<Expr*> uses_expr = this->fusion()->unordered_uses(this);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: do we need to make a copy of the uses? how about 
`const std::unordered_set<Expr*>& uses_expr = this->fusion()->unordered_uses(this);`

or
`const auto& uses_expr = this->fusion()->unordered_uses(this);`

... or even better:
`for (auto expr : usion()->unordered_uses(this);) { ... } `

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 14]
```diff
@@ -314,6 +314,331 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
   return producer;
 }
 
+// Create a TensorView before the original tensor. A common use case is to write
+// results into shared memory or local registers before moving to global memory.
+// Analogous to TVM Cache_Write
+TensorView* TensorView::cache_before() {
+  FusionGuard fg(this->fusion());
+
+  Expr* origin_expr = this->fusion()->origin(this);
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: no need for `this->`. There are multiple instances of this pattern, please do a search and replace.

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 15]
```diff
@@ -314,6 +314,331 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
...
+
+    auto rel_ca_pos = this->getRelativeComputeAtAxis();
+    auto this_ca_pos = this->getThisComputeAtAxis();
+    auto this_ca_view = this->getComputeAtView();
+
+    this->computeAt(consumer, this_ca_pos);
+    consumer->setComputeAt(this_ca_view, rel_ca_pos);
+  } else {
+    // Check users of this TV for computeAt for cache_after on inputs
+    std::unordered_set<Expr*> consumer_uses =
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: ditto

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 16]
```diff
@@ -314,6 +314,331 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
...
+        consumer_,
+        reduction_expr->in());
+  }
+
+  void handle(BroadcastOp* broadcast_expr) final {
+    new BroadcastOp(consumer_, broadcast_expr->in());
+  }
+
+ private:
+  TensorView* consumer_;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: ` = nullptr;`

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 17]
```diff
@@ -314,6 +314,331 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
...
+        reduction_expr->out(),
+        producer_);
+  }
+
+  void handle(BroadcastOp* broadcast_expr) final {
+    new BroadcastOp(broadcast_expr->out(), producer_);
+  }
+
+ private:
+  TensorView* current_;
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: = nullptr;

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 18]
```diff
@@ -314,6 +314,331 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
...
+  TensorView* producer_;
+};
+} // namespace
```
Comment (tlemo) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp:622]: nit: empty line before }

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp (hunk 11 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp:1030]: Naoya found an issue with `fusion->printKernel` and the cache functions. It was adding an extra output to the TensorView.

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 13 above)
Comment (rdspring1) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp]: `for (auto expr : fusion()->unordered_uses(this);) { ... }` is fine.

File: torch/csrc/jit/codegen/cuda/ir_nodes.cpp (hunk 11 above)
Comment (tlemo) [reply torch/csrc/jit/codegen/cuda/ir_nodes.cpp:1030]: thanks

**Review by naoyam — APPROVED** [review]
LGTM. Please update the document regarding the limitation and transformation examples.
